### ISSUE TITLE
Functional units in FITS and VOTable

### ISSUE THREAD

**Issue by mhvk** [astropy/astropy#3822]
This is to remind me/us that now that function units are present (#1894), it should also be possible to write/read them to FITS and VOTable, at least for the subset for which the standard allows this.

The general writer is addressed in #3618.

**Comment by embray** [comment]
:+1:

**Comment by pllim** [comment]
I removed the milestone, @mhvk . FYI.

**Comment by pllim** [comment]
@olebole would also like a new `u.log` unit. See https://github.com/astropy/astropy/issues/16540#issuecomment-2150676682

**Comment by olebole** [comment]
I'd like to bring this back to the top again, because it seems to me the issue is almost complete now:

* we already can generally parse functional units, see #3554
* we can generally write functional units

This should be rather trivial, not? I could try to create a PR here, but before I would like to clarify what prevented us from implementing this also for FITS and VO, at least for dimensionless **mag** and for **dex** (which can be represented with the `log` function in FITS and VO)?

i.e.
```Python
>>> u.mag(1).to_string("vounit")
'mag'
>>> u.dex(1).to_string("vounit")
'log(1)'
>>> u.dex(u.solLum).to_string("vounit")
'log(solLum)'
```

The first one doesn't pass a roundtrip, as `u.Unit("mag", format="vounit")` doesn't return a **magUnit**, but the log/dex representation should be simple, not?

**Comment by pllim** [comment]
Also cc @eerovaher and @nstarman who also work on `units` a lot lately.

**Comment by nstarman** [comment]
> I would like to clarify what prevented us from implementing this also for FITS and VO, at least for dimensionless mag and for dex 

I think that was before my time.
Seems like something we should support, and if it's easy to do now, that's great!

**Comment by mhvk** [comment]
It never was particularly difficult, just needs someone to put in the time to think how exactly it should work and then implement it -- apparently, I didn't find that time in 9 years... If someone can write a series of tests, that would be a great start...

**Comment by jwfraustro** [comment]
This is something I have just bumped into as well, related to the above https://github.com/astropy/astropy/issues/17297. I would love to see this as well.

In our case, it was from another package that was using `astropy.units` to validate FITS and VO unit strings. I was very surprised when I found that `VOUnit("log(solMass)")` did not pass as valid, when the syntax is allowed by the standard.

**Comment by mhvk** [comment]
Yes, it would be good. Not sure I'll get around it myself soon, but there are some hints here and especially in #17297 on how to implement it. As I wrote here, even just having a good set of tests would be a great start.

**Comment by gpdf** [comment]
Perhaps the original issue was that log/ln/exp() function units are mathematically (very) different from "ordinary" units, which are inherently multiplicative.  Note that they are not at all like sqrt() -- e.g., `sqrt(m)` is still a multiplicative unit.

Is the issue not one of string-parsing, but rather one of what Astropy is supposed to do when a unit conversion is requested?  Perhaps that code isn't ready for this?

**Comment by mhvk** [comment]
@gpdf - astropy supports these "functional" units, which are used for `Magnitude`, `Decibel`, etc. The issue here is just about correctly translating these to FITS and VOTable (as is already possible for CDS format).

**Comment by gpdf** [comment]
@jwfraustro Yes, understood.  So perhaps the situation is just "haven't gotten to it" rather than "difficult to implement correctly"?
